triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/ServiceTest.scala

```diff
@@ -331,4 +325,50 @@ class ServiceTest extends AsyncFlatSpec with Eventually with Matchers with Postg
       ledgerProxy.enable()
     }
   }
+
+  it should "stop a failing trigger that can't be restarted" in withHttpService(Some(dar)) {
+    (uri: Uri, client: LedgerClient, ledgerProxy: Proxy) =>
+      // Simulate the ledger becoming unrecoverably unavailable due to
+      // network connectivity loss. The stop strategy means the running
+      // trigger will be terminated.
+      for {
+        // Request a trigger be started for Alice.
+        resp <- startTrigger(uri, s"$testPkgId:TestTrigger:trigger", "Alice")
+        aliceTrigger <- parseTriggerId(resp)
+        // Proceed when it's confirmed to be running.
+        _ <- assertTriggerIds(uri, "Alice", _ == Vector(aliceTrigger))
+        // Simulate unrecoverable network connectivity loss.
+        _ <- Future { ledgerProxy.disable() }
+        // Confirm that the running trigger is stopped.
+        _ <- assertTriggerIds(uri, "Alice", _.isEmpty)
+      } yield succeed
+  }
+
+  it should "restart a failing trigger if possible" in withHttpService(Some(dar)) {
+    (uri: Uri, client: LedgerClient, ledgerProxy: Proxy) =>
+      // Simulate the ledger briefly being unavailable due to network
+      // connectivity loss. Our restart strategy means that the running
+      // trigger gets restarted.
+      for {
+        // Request a trigger be started for Alice.
+        resp <- startTrigger(uri, s"$testPkgId:TestTrigger:trigger", "Alice")
+        aliceTrigger <- parseTriggerId(resp)
+        // Proceed when it is confirmed to be running.
+        _ <- assertTriggerIds(uri, "Alice", _ == Vector(aliceTrigger))
+        // Simulate brief network connectivity loss. This will cause the
+        // running trigger's flow graph to complete with failure. Don't
+        // wait around to restore the network or the restart strategy
+        // will in turn lead to the stop strategy killing the trigger
+        // due to the lack of ability to initialize the restarted
+        // trigger.
+        _ <- Future { ledgerProxy.disable() }
```

Comment:
Is this timing sensitive because we only retry a limited number of times? I guess if we add the `assertTriggerIds(uri, "Alice", _.isEmpty)` part from the other test it would fail since we’ve reached the maximum number of retries?

Reply:
Recall there are two supervisor strategies. Failures to initialize result in stopped triggers, execution failures result in restarts. Suppose an execution failure occurs due to network outage - an attempt will be made to restart it. If the outage is still present as the restart cycles through initialization then the trigger will get stopped (permanently) as an initialization failure. That's where this timing sensitivity comes from. I think this answer indicates that there's room for improvement but I'm not clear yet on how we might do so. To the second question, yes, the wait involved in the `assertTriggerIds` means we'll not recover the network in time for the restart to succeed but as indicated above, not because the max retries of the trigger have been reached but rather because it will be stopped due to initialization failure.